google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb

```diff
@@ -0,0 +1,566 @@
+                      response.target_change.read_time
+                    )
+                  )
+
+                  if @inventory.current? && @inventory.changes?
+                    synchronize do
+                      send_callback @inventory.build_query_snapshot
+                    end
+                  end
+                when :CURRENT
+                  # The targets reflect all changes committed before the targets
+                  # were added to the stream.
+                  #
+                  # This will be sent after or with a +read_time+ that is
+                  # greater than or equal to the time at which the targets were
+                  # added.
+                  #
+                  # Listeners can wait for this change if read-after-write
+                  # semantics are desired.
+
+                  @inventory.persist(
+                    response.target_change.resume_token,
+                    Convert.timestamp_to_time(
+                      response.target_change.read_time
+                    )
+                  )
+
+                  @inventory.current = true
+                when :RESET
+                  # The targets have been reset, and a new initial state for the
+                  # targets will be returned in subsequent changes.
+                  #
+                  # After the initial state is complete, +CURRENT+ will be
+                  # returned even if the target was previously indicated to be
+                  # +CURRENT+.
+
+                  raise RestartStream # Raise to restart the stream
+                end
+              when :document_change
+                # A {Google::Firestore::V1beta1::Document Document} has changed.
+
+                if response.document_change.removed_target_ids.any?
+                  @inventory.delete response.document_change.document.name
+                else
+                  @inventory.add response.document_change.document
+                end
+              when :document_delete
+                # A {Google::Firestore::V1beta1::Document Document} has been
+                # deleted.
+
+                @inventory.delete response.document_delete.document
+              when :document_remove
+                # A {Google::Firestore::V1beta1::Document Document} has been
+                # removed from a target (because it is no longer relevant to
+                # that target).
+
+                @inventory.delete response.document_remove.document
+              end
+            rescue StopIteration
+              break
+            end
+
+            # Reset backoff values when completed without an error
+            @backoff[:current] = 0
+            @backoff[:delay] = 1.0
+          end
+
+          # Has the loop broken but we aren't stopped?
+          # Could be GRPC has thrown an internal error, so restart.
+          raise RestartStream
+        rescue GRPC::DeadlineExceeded, GRPC::Unavailable, GRPC::Cancelled,
```

Comment:
@schmidt-sebastian Can you verify the errors to retry with incremental backoff list here is correct? This is the same list that Spanner is using to restart streams with.

Reply:
This is not the list we use in the other clients, but I am generally ok with keeping this consistent with Spanner.  The one exception to this is `UNAUTHENTICATED`. Please verify that you can re-open the stream when the authentication expires. This will happen after 60 minutes.

Reply:
I’ll add UNAUTHENTICATED.

Reply:
This has been added.